spec/lib/submission_builder/ty2024/states/nj/nj_return_xml_spec.rb

```diff
@@ -13,5 +13,85 @@
       expect(xml.document.at('AuthenticationHeader').to_s).to include('xmlns="http://www.irs.gov/efile"')
       expect(xml.document.at('ReturnHeaderState').to_s).to include('xmlns="http://www.irs.gov/efile"')
     end
+
+    context "when filer has no spouse" do
+      it "Exemptions are populated" do
+        expect(xml.css('Exemptions').count).to eq(1)
+      end
+
+      it "populates line 6 XML fields" do
+        expect(xml.at("Exemptions YouRegular").text).to eq("X")
+        expect(xml.at("Exemptions SpouseCURegular").text).to eq("")
+        expect(xml.at("Exemptions DomesticPartnerRegular").text).to eq("")
+        expect(xml.at("Exemptions Line6").text).to eq("1000")
+      end
+
+      context "when filer is over 65" do
+        let(:intake) { create(:state_file_nj_intake, :self_over_65) }
+        it "populates line 7 XML fields" do
+          expect(xml.at("Exemptions YouOver65").text).to eq("X")
+          expect(xml.at("Exemptions SpouseCuPartner65OrOver").text).to eq("")
+          expect(xml.at("Exemptions Line7").text).to eq("1000")
+        end
+      end
+      context "when filer is younger than 65" do
+        let(:intake) { create(:state_file_nj_intake) }
+        it "populates line 7 XML fields" do
+          expect(xml.at("Exemptions YouOver65").text).to eq("")
+          expect(xml.at("Exemptions SpouseCuPartner65OrOver").text).to eq("")
+          expect(xml.at("Exemptions Line7").text).to eq("0")
+        end
+      end
+    end
+
+    context "when filer is married" do
+      let(:intake) { create(:state_file_nj_intake, :married) }
+      it "Exemptions are populated" do
+        expect(xml.css('Exemptions').count).to eq(1)
+      end
+
+      it "populates line 6 XML fields" do
+        expect(xml.at("Exemptions YouRegular").text).to eq("X")
+        expect(xml.at("Exemptions SpouseCURegular").text).to eq("X")
+        expect(xml.at("Exemptions DomesticPartnerRegular").text).to eq("")
+        expect(xml.at("Exemptions Line6").text).to eq("2000")
+      end
+
+      context "when filer is over 65 and spouse is under 65" do
+        let(:intake) { create(:state_file_nj_intake, :self_over_65, :married) }
+        it "populates line 7 XML fields" do
+          expect(xml.at("Exemptions YouOver65").text).to eq("X")
+          expect(xml.at("Exemptions SpouseCuPartner65OrOver").text).to eq("")
+          expect(xml.at("Exemptions Line7").text).to eq("1000")
+        end
+      end
+
+      context "when filer is over 65 and spouse is over 65" do
```

Comment:
Is there a better way to dry up these tests for every possible test-case? Concerned that I'm already adding more tests for just these two lines as the length of the tests in the entire `az140_calculator spec`.

It seems like I'm duplicating a lot of logic between the XML generation and the calculator, also testing the cases where no spouse exists separately from the case where a spouse exists and is under/over 65.

Reply:
Same kind of feedback as the the other spec test. I think simplifying and nesting contexts that are similar into each other ("when filer is over 65 and spouse is over 65" could be when spouse is over 65 in the married context instead) and encompassing "it" statements for multiple lines could make it more concise. We also don't necessarily do tests for every matrix of possibility although it is theoretically good to do that and up to your discretion. For example, IMO I would probably just test for spouse 65 and older and younger and not the combination of filer and spouse both 65 and older or one 65 or older.